fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.tool.attribute;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.tag.Tag;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.TypedActionResult;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.tool.attribute.v1.DynamicAttributeTool;
+import net.fabricmc.fabric.api.util.TriState;
+
+public final class ToolManagerImpl {
+	public interface Entry {
+		void setBreakByHand(boolean value);
+
+		void putBreakByTool(Tag<Item> tag, int miningLevel);
+
+		int getMiningLevel(Tag<Item> tag);
+	}
+
+	private static class EntryImpl implements Entry {
+		private Tag<Item>[] tags = new Tag[0];
+		private int[] tagLevels = new int[0];
+		private TriState defaultValue = TriState.DEFAULT;
+
+		@Override
+		public void setBreakByHand(boolean value) {
+			this.defaultValue = TriState.of(value);
+		}
+
+		@Override
+		public void putBreakByTool(Tag<Item> tag, int miningLevel) {
+			tag(tag); // Generate tag entry
+
+			for (int i = 0; i < tags.length; i++) {
+				if (tags[i] == tag) {
+					tagLevels[i] = miningLevel;
+					return;
+				}
+			}
+
+			Tag<Item>[] newTags = new Tag[tags.length + 1];
+			int[] newTagLevels = new int[tagLevels.length + 1];
+			System.arraycopy(tags, 0, newTags, 0, tags.length);
+			System.arraycopy(tagLevels, 0, newTagLevels, 0, tagLevels.length);
```

Comment:
I assume these tags you array copy are either already generated or null checked?

Reply:
This was from the old tool manager, I renamed the class and now it shows up like this, I _might_ want to move these to lists but I don't have an idea on the performance impact.

Reply:
well hmm nevermind, it seems like I using arrays might not be a bad idea

Reply:
Should use Arrays.copyOf though